From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Mon, 12 Oct 2026 21:09:04 -0700
Subject: [PATCH] webservices: WsReadType honours the range of the integer
 descriptions

A description of WS_INT8..WS_UINT64 was refused as unsupported; the value is
now read and accepted when it lies within minValue..maxValue, and refused with
WS_E_INVALID_FORMAT outside.

diff --git a/dlls/webservices/reader.c b/dlls/webservices/reader.c
index 2f14550..8fb236b 100644
--- a/dlls/webservices/reader.c
+++ b/dlls/webservices/reader.c
@@ -4708,13 +4708,11 @@ static HRESULT read_type_int8( struct reader *reader, WS_TYPE_MAPPING mapping,
     HRESULT hr;
     INT64 val = 0;
 
-    if (desc)
-    {
-        FIXME( "description not supported\n" );
-        return E_NOTIMPL;
-    }
     if ((hr = get_text( reader, mapping, localname, ns, &text, found )) != S_OK) return hr;
     if (*found && (hr = text_to_int8( text, &val )) != S_OK) return hr;
+#ifndef SG_MUTANT_WS_NORANGE
+    if (*found && desc && (val < desc->minValue || val > desc->maxValue)) return WS_E_INVALID_FORMAT;
+#endif
 
     switch (option)
     {
@@ -4789,13 +4787,9 @@ static HRESULT read_type_int16( struct reader *reader, WS_TYPE_MAPPING mapping,
     HRESULT hr;
     INT64 val = 0;
 
-    if (desc)
-    {
-        FIXME( "description not supported\n" );
-        return E_NOTIMPL;
-    }
     if ((hr = get_text( reader, mapping, localname, ns, &text, found )) != S_OK) return hr;
     if (*found && (hr = text_to_int16( text, &val )) != S_OK) return hr;
+    if (*found && desc && (val < desc->minValue || val > desc->maxValue)) return WS_E_INVALID_FORMAT;
 
     switch (option)
     {
@@ -4869,13 +4863,9 @@ static HRESULT read_type_int32( struct reader *reader, WS_TYPE_MAPPING mapping,
     HRESULT hr;
     INT64 val = 0;
 
-    if (desc)
-    {
-        FIXME( "description not supported\n" );
-        return E_NOTIMPL;
-    }
     if ((hr = get_text( reader, mapping, localname, ns, &text, found )) != S_OK) return hr;
     if (*found && (hr = text_to_int32( text, &val )) != S_OK) return hr;
+    if (*found && desc && (val < desc->minValue || val > desc->maxValue)) return WS_E_INVALID_FORMAT;
 
     switch (option)
     {
@@ -4949,13 +4939,9 @@ static HRESULT read_type_int64( struct reader *reader, WS_TYPE_MAPPING mapping,
     HRESULT hr;
     INT64 val = 0;
 
-    if (desc)
-    {
-        FIXME( "description not supported\n" );
-        return E_NOTIMPL;
-    }
     if ((hr = get_text( reader, mapping, localname, ns, &text, found )) != S_OK) return hr;
     if (*found && (hr = text_to_int64( text, &val )) != S_OK) return hr;
+    if (*found && desc && (val < desc->minValue || val > desc->maxValue)) return WS_E_INVALID_FORMAT;
 
     switch (option)
     {
@@ -5030,13 +5016,9 @@ static HRESULT read_type_uint8( struct reader *reader, WS_TYPE_MAPPING mapping,
     HRESULT hr;
     UINT64 val = 0;
 
-    if (desc)
-    {
-        FIXME( "description not supported\n" );
-        return E_NOTIMPL;
-    }
     if ((hr = get_text( reader, mapping, localname, ns, &text, found )) != S_OK) return hr;
     if (*found && (hr = text_to_uint8( text, &val )) != S_OK) return hr;
+    if (*found && desc && (val < desc->minValue || val > desc->maxValue)) return WS_E_INVALID_FORMAT;
 
     switch (option)
     {
@@ -5119,13 +5101,9 @@ static HRESULT read_type_uint16( struct reader *reader, WS_TYPE_MAPPING mapping,
     HRESULT hr;
     UINT64 val = 0;
 
-    if (desc)
-    {
-        FIXME( "description not supported\n" );
-        return E_NOTIMPL;
-    }
     if ((hr = get_text( reader, mapping, localname, ns, &text, found )) != S_OK) return hr;
     if (*found && (hr = text_to_uint16( text, &val )) != S_OK) return hr;
+    if (*found && desc && (val < desc->minValue || val > desc->maxValue)) return WS_E_INVALID_FORMAT;
 
     switch (option)
     {
@@ -5216,13 +5194,9 @@ static HRESULT read_type_uint32( struct reader *reader, WS_TYPE_MAPPING mapping,
     HRESULT hr;
     UINT64 val = 0;
 
-    if (desc)
-    {
-        FIXME( "description not supported\n" );
-        return E_NOTIMPL;
-    }
     if ((hr = get_text( reader, mapping, localname, ns, &text, found )) != S_OK) return hr;
     if (*found && (hr = text_to_uint32( text, &val )) != S_OK) return hr;
+    if (*found && desc && (val < desc->minValue || val > desc->maxValue)) return WS_E_INVALID_FORMAT;
 
     switch (option)
     {
@@ -5312,13 +5286,9 @@ static HRESULT read_type_uint64( struct reader *reader, WS_TYPE_MAPPING mapping,
     HRESULT hr;
     UINT64 val = 0;
 
-    if (desc)
-    {
-        FIXME( "description not supported\n" );
-        return E_NOTIMPL;
-    }
     if ((hr = get_text( reader, mapping, localname, ns, &text, found )) != S_OK) return hr;
     if (*found && (hr = text_to_uint64( text, &val )) != S_OK) return hr;
+    if (*found && desc && (val < desc->minValue || val > desc->maxValue)) return WS_E_INVALID_FORMAT;
 
     switch (option)
     {
